x/dualstaking/keeper/delegator_reward.go

```diff
@@ -0,0 +1,97 @@
+
+	b := store.Get(types.DelegatorRewardKey(
+		index,
+	))
+	if b == nil {
+		return val, false
+	}
+
+	k.cdc.MustUnmarshal(b, &val)
+	return val, true
+}
+
+// RemoveDelegatorReward removes a DelegatorReward from the store
+func (k Keeper) RemoveDelegatorReward(
+	ctx sdk.Context,
+	index string,
+) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.DelegatorRewardKeyPrefix))
+	store.Delete(types.DelegatorRewardKey(
+		index,
+	))
+}
+
+// GetAllDelegatorReward returns all DelegatorReward
+func (k Keeper) GetAllDelegatorReward(ctx sdk.Context) (list []types.DelegatorReward) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.DelegatorRewardKeyPrefix))
+	iterator := sdk.KVStorePrefixIterator(store, []byte{})
+
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var val types.DelegatorReward
+		k.cdc.MustUnmarshal(iterator.Value(), &val)
+		list = append(list, val)
+	}
+
+	return
+}
+
+// CalcProviderReward calculates the provider reward considering delegations
+// providerReward = totalReward * ((effectiveDelegations*commission + providerStake) / effective stake)
+func (k Keeper) CalcProviderReward(stakeEntry epochstoragetypes.StakeEntry, totalReward math.Int) math.Int {
+	providerStake := stakeEntry.Stake.Amount
+	delegationCommission := stakeEntry.DelegateCommission
+	effectiveDelegations, effectiveStake := k.CalcEffectiveDelegationsAndStake(stakeEntry)
+
+	providerRewardPercentage := effectiveDelegations.MulRaw(int64(delegationCommission / 100)).Add(providerStake).Quo(effectiveStake)
+	return providerRewardPercentage.Mul(totalReward)
+}
+
+// CalcDelegatorsReward calculates the total amount of rewards for all delegators
+// delegatorsReward = totalReward - providerReward
+func (k Keeper) CalcDelegatorsReward(stakeEntry epochstoragetypes.StakeEntry, totalReward math.Int) math.Int {
+	return totalReward.Sub(k.CalcProviderReward(stakeEntry, totalReward))
+}
+
+// CalcEffectiveDelegationsAndStake calculates the effective stake and effective delegations (for delegator rewards calculations)
+// effectiveDelegations = totalDelegations
+// effective stake = effectiveDelegations + providerStake
+func (k Keeper) CalcEffectiveDelegationsAndStake(stakeEntry epochstoragetypes.StakeEntry) (effectiveDelegations math.Int, effectiveStake math.Int) {
+	totalDelegations := stakeEntry.DelegateTotal.Amount
+	return totalDelegations, totalDelegations.Add(stakeEntry.Stake.Amount)
+}
+
+// CalcDelegatorReward calculates a single delegator reward according to its delegation
+// delegatorReward = (delegatorsReward * delegatorStake) / totalDelegations
+func (k Keeper) CalcDelegatorReward(stakeEntry epochstoragetypes.StakeEntry, totalReward math.Int, delegation types.Delegation) math.Int {
+	totalDelegations := stakeEntry.DelegateTotal.Amount
+
+	delegatorsReward := k.CalcDelegatorsReward(stakeEntry, totalReward)
```

Comment:
this method is called in a for loop but always has the same result, call it before the loop once

Reply:
done. see https://github.com/lavanet/lava/commit/4985b44d36d626bb5dcb72cd8a1fe821e723a21a